test/src/kbasesearchengine/test/events/handler/WorkspaceEventHandlerTest.java

```diff
@@ -1514,7 +1514,7 @@ public void newVersionPublicFlagUpdatedFromWorkspaceCase1() throws Exception {
                 // for name check
                 .thenReturn(new UObject(new GetObjectInfo3Results().withInfos(objList)))
                 // for isPublic check
-                .thenReturn(new UObject(wsTuple(3L, "wsname", "username", "date", 7, "n", "r",
+                .thenReturn(new UObject(wsTuple(3L, "wsname", "username", "date", 7, "n", new String("r"),
```

Comment:
why new String

Reply:
This is tied to the fix to replace == for string comparison with equals. When using == two strings are only equal if they are the same string object or are interned strings (e.g. derived form the same string literal or using new String("foo").intern()). When you say "foo" the string literals are converted to "interned" string objects -- a special pool of string objects in the JVM for which there is only one object for each unique string. Each of the "foo"s in "foo" == "foo" resolves to the same interned string object, and so it evaluates to true. But when you compare non-interned strings, even if the string is the same, == gives you false, since they are different objects. In all cases equals works as expected.

The problem with using string literals in the test in this case is that in the actual workspace api, the string values returned are always (at present) non-interned string objects. Thus, if you get the public permission value from a workspace api call, and compare it to some other value  some string value with == it will always be false. However, the workspace mock returns exactly what you give to the "thenReturn" method, which in this case was a string literal / interned string object.

The WorkspaceEventHandler.getLatestIsPublic, which determines the current state of the public permission for the workspace, uses a string literal "n" for comparison.

So the pre-fixed code in WorkspaceEventHandler.getLatestIsPublic used == and would succeed in the test, but fail in real life when a real workspace call was made. 

So, replacing == with equals fixed that problem.

On the other hand, using "n" for the mock is wrong because the non-mocked api returns a non-interned string.

So, many times more words than required, the usage of new String("r") is more correct than "r" for the current behavior of the workspace api. (Though admittedly the workspace api could return interned strings at some point in the future.) However, after replacing == with equals in WorkspaceEventHandler.getLatestIsPublic it doesn't make a difference.

In my opinion, the mock should accurately reflect the precise nature of the real api in this respect, because otherwise it can lead test results to differ from real-life results in subtle ways. Imagine someone in the future decides to replace equals with == in getLatestIsPublic, and feels proud because it is more efficient and the tests still pass. It would fail in a real deployment.

In this case I only "fixed" the one field of the get_workspace_info simulated result, since it is the only one used in this test which would be affected.

Reply:
Something like that, I'm no java pro, though.